### DISCUSSION TITLE
How to define custom TMSs with variableMatrixWidths?

### DISCUSSION THREAD

**Discussion by AndrewAnnex** [developmentseed/morecantile#155] [Q&A] [answered]
I was looking again at #131 and I don't see an obvious way within morecantile for constructing custom tile matrix sets with variable matrix widths programatically. I suppose I could always clone the dict for GNOSISGlobalGrid and update values for my particular needs and do something like what is done in tests/test_tms_variable_width.py to define the test TMS, but I think there would be a way to make this behavior configurable for the .custom method.

**Comment by AndrewAnnex** [comment]
actually maybe that test does in fact show that setting `variableMatrixWidths` is all that is needed

**Reply by vincentsarago** [reply]
there is not `easy` way to create such TMS (like the `.custom` method so you'll need to create all the matrices manually has we do in https://github.com/developmentseed/morecantile/blob/main/tests/test_tms_variable_width.py#L16-L34

**Reply by AndrewAnnex** [reply, marked as answer]
I ended up doing some work that seems to reproduce the gnosis global grid and wrote some functions that seem to do the job:

```python
import numpy as np 

def coalesing_coefficients(z):
    bottom = 2 ** np.arange(1, z+1)
    return np.concatenate((bottom[::-1], bottom))

def coalesing_tile_rows(z):
    max_row = 2 ** (z + 1) - 1
    top_min_rows = np.concatenate(([0], 2 ** np.arange(z-1)))
    top_max_rows = np.cumsum(top_min_rows)
    bot_max_rows = max_row - top_min_rows
    bot_min_rows = max_row - top_max_rows
    min_rows = np.hstack((top_min_rows, bot_min_rows[::-1]))
    max_rows = np.hstack((top_max_rows, bot_max_rows[::-1]))
    return min_rows, max_rows

def get_variable_matrix_widths(z):
    coefficients = coalesing_coefficients(z)
    if len(coefficients) == 0:
        return []
    min_rows, max_rows = coalesing_tile_rows(z)
    for c, min_r, max_r in zip(coefficients, min_rows, max_rows):
        yield dict(coalesce = c.item(), minTileRow=min_r.item(), maxTileRow=max_r.item())

```
